### PR TITLE
Fold IND(LDELEMA_REF(obj, index, type)) to INDEX(obj, index)

### LINKED ISSUE #65789: Unnecessary LdelemaRef call not elided with Volatile.Read
Consider the code:
```C#
using System;
using System.Threading;

public sealed class C
{
    private readonly C[] _array;
    
    public C Get1() => Volatile.Read(ref _array[0]);
    
    public C Get2() => _array[0];
}
```
We shouldn't need a helper to read a value from the array, but even though the `Volatile.Read` evaporates on x86/64, we're still left with a non-inlined CastHelpers.LdelemaRef call in Get1:
```
; Core CLR 6.0.121.56705 on x86

C.Get1()
    L0000: mov ecx, [ecx+4]
    L0003: push 0x1171c6d8
    L0008: xor edx, edx
    L000a: call System.Runtime.CompilerServices.CastHelpers.LdelemaRef(System.Array, Int32, Void*)
    L000f: mov eax, [eax]
    L0011: ret

C.Get2()
    L0000: mov eax, [ecx+4]
    L0003: cmp dword ptr [eax+4], 0
    L0007: jbe short L000d
    L0009: mov eax, [eax+8]
    L000c: ret
    L000d: call 0x722c5fc0
    L0012: int3

; Core CLR 6.0.121.56705 on amd64

C.Get1()
    L0000: sub rsp, 0x28
    L0004: mov rcx, [rcx+8]
    L0008: xor edx, edx
    L000a: mov r8, 0x7ffaab0fccd0
    L0014: call 0x00007ffa9fda8200
    L0019: mov rax, [rax]
    L001c: add rsp, 0x28
    L0020: ret

C.Get2()
    L0000: sub rsp, 0x28
    L0004: mov rax, [rcx+8]
    L0008: cmp dword ptr [rax+8], 0
    L000c: jbe short L0017
    L000e: mov rax, [rax+0x10]
    L0012: add rsp, 0x28
    L0016: ret
    L0017: call 0x00007ffaffa6ec10
    L001c: int3
```
[SharpLab](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSB0AKgBYBOApmACYCWAdgOYDc66cAzNhJQDblW4E2AMLoA3umxTsAB1I0AbmGDlsFagHs6PAJ4iA2gF1sAfTClSYHSzTTsk6RxHYA4uWBIAFAEpsAXgA+bAA1DR5lGj58ACVKKk8KADNTc0sdfUxDbxs7ByknYVd3BB9/ILMLKwzDGwBfIA=)

category:cq
theme:volatile

### PR DESCRIPTION
**Pull request by EgorBo**
Closes https://github.com/dotnet/runtime/issues/65789 let's see what diffs think
```csharp
public sealed class C
{
    C[] _array;

    C Test1() => Volatile.Read(ref _array[3]);

    void Test2()
    {
        ref C c = ref _array[0];
        c = null;
    }
}
```
Codegen diff: https://www.diffchecker.com/VQ8cweA8

### CODE REVIEW

File: src/coreclr/jit/morph.cpp [hunk 1]
```diff
@@ -11541,6 +11541,37 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
...
+            {
+                GenTreeCall* call = op1->AsCall();
+                if ((call->gtCallType == CT_HELPER) &&
+                    (eeGetHelperNum(call->gtCallMethHnd) == CORINFO_HELP_LDELEMA_REF))
+                {
+                    GenTree*             objOp   = call->gtCallArgs->GetNode();
+                    GenTree*             indexOp = call->gtCallArgs->GetNext()->GetNode();
+                    GenTree*             clsOp   = call->gtCallArgs->GetNext()->GetNext()->GetNode();
+                    CORINFO_CLASS_HANDLE cls     = gtGetHelperArgClassHandle(clsOp);
+                    if ((cls != NO_CLASS_HANDLE) && impIsClassExact(cls))
```
Comment (jakobbotsch) [inline src/coreclr/jit/morph.cpp:11557]: Doesn't this need to check if the array element type is exact and equal to `cls` as well?

Comment (EgorBo) [reply src/coreclr/jit/morph.cpp:11557]: Yes, not sure it's possible that they won't match in pure C#, but sure, I'll add

Comment (jakobbotsch) [reply src/coreclr/jit/morph.cpp:11557]: Isn't it necessary to handle covariance, even for C#?

Comment (EgorBo) [reply src/coreclr/jit/morph.cpp:11557]: > Isn't it necessary to handle covariance, even for C#?

not sure I understand what you mean - do you have an example where `ldelema's type is X (and it's sealed)` but the actual array is something different? (in C#)

Comment (jakobbotsch) [reply src/coreclr/jit/morph.cpp]: Ah, I misunderstood what `impIsClassExact` was checking, looks fine to me now.

Comment (EgorBo) [reply src/coreclr/jit/morph.cpp]: well, it worth checking types anyway so thanks for the suggestion 🙂

**Comment by EgorBo** [comment]
hm... no diffs, @stephentoub did you expect it to improve some already existing codepath or a planned one only?

**Comment by stephentoub** [comment]
Yes, e.g. https://github.com/dotnet/runtime/blob/9212310d20689b3ac7213569719b0f50a2274f5a/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs#L402

**Comment by EgorBo** [comment]
Ah, I see. Unfortunately that case is not handled by "Forward Substitution" pass which is quite conservative at the moment and I believe it explains zero diffs in general, so instead of IND(LDELEMA_REF) I see ASG(LDELEMA_REF)

I'm going to close it and, instead, try to make all CastHelpers methods inlineable for hot paths so e.g. ldelemaRef would be inlineable - it should improve perf for that case

**Pull request closed without merging**